plugins/catalog-backend/src/ingestion/processors/GitLabDiscoveryProcessor.ts

```diff
@@ -0,0 +1,174 @@
+    if (!integration) {
+      throw new Error(
+        `There is no GitLab integration that matches ${host}. Please add a configuration entry for it under integrations.gitlab`,
+      );
+    }
+
+    const client = new GitLabClient({
+      config: integration.config,
+      logger: this.logger,
+    });
+    const startTimestamp = Date.now();
+    this.logger.info(`Reading GitLab projects from ${location.target}`);
+
+    const projects = paginated(options => client.listProjects(options), {
+      group,
+      last_activity_after: await this.updateLastActivity(),
+      page: 1,
+    });
+
+    const result: Result = {
+      scanned: 0,
+      matches: [],
+    };
+    for await (const project of projects) {
+      result.scanned++;
+      if (!project.archived) {
+        result.matches.push(project);
+      }
+    }
+
+    for (const project of result.matches) {
+      const project_branch = branch === '*' ? project.default_branch : branch;
+
+      emit(
+        results.location(
+          {
+            type: 'url',
+            // The format expected by the GitLabUrlReader:
+            // https://gitlab.com/groupA/teams/teamA/subgroupA/repoA/-/blob/branch/filepath
+            //
+            // This unfortunately will trigger another API call in `getGitLabFileFetchUrl` to get the project ID.
+            // The alternative is using the `buildRawUrl` function, which does not support subgroups, so providing a raw
+            // URL here won't work either.
+            target: `${project.web_url}/-/blob/${project_branch}/${catalogPath}`,
+          },
+          true,
+        ),
+      );
+    }
+
+    const duration = ((Date.now() - startTimestamp) / 1000).toFixed(1);
+    this.logger.debug(
+      `Read ${result.scanned} GitLab repositories in ${duration} seconds`,
+    );
+
+    return true;
+  }
+
+  async updateLastActivity(): Promise<string | undefined> {
+    const lastActivity = await this.cache.get('last-activity');
+    await this.cache.set('last-activity', new Date().toISOString());
+    return lastActivity as string | undefined;
+  }
+}
+
+type Result = {
+  scanned: number;
+  matches: GitLabProject[];
+};
+
+/*
+ * Helpers
+ */
+
+export function parseUrl(urlString: string): {
+  group?: string;
+  host: string;
+  branch: string;
+  catalogPath: string;
+} {
+  const url = new URL(urlString);
+  const path = url.pathname.substr(1).split('/');
+
+  // (/group/subgroup)/blob/branch|*/filepath
+  const blobIndex = path.findIndex(p => p === 'blob');
+  if (blobIndex !== -1 && path.length > blobIndex + 2) {
+    const group =
+      blobIndex > 0 ? path.slice(0, blobIndex).join('/') : undefined;
+
+    return {
+      group,
+      host: url.host,
+      branch: decodeURIComponent(path[blobIndex + 1]),
+      catalogPath: decodeURIComponent(path.slice(blobIndex + 2).join('/')),
+    };
+  }
+
+  throw new Error(`Failed to parse ${urlString}`);
+}
+
+export function escapeRegExp(str: string): RegExp {
```

Comment:
It does not look like this function is used for anything?

Reply:
Whoops, that got left over from a copy/paste, I'll remove it.